mempool/mempool-task/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+    }
+
+    fn on_blockchain_event(&mut self, event: &BlockchainEvent) {
+        match event {
+            BlockchainEvent::HistoryAdopted(_) => {
+                // Mempool updates are only done once we are synced.
+                if self.consensus.is_ready_for_validation() {
+                    self.mempool.cleanup();
+                    debug!("Performed a mempool clean up because new history was adopted");
+                }
+            }
+            BlockchainEvent::Extended(hash) => {
+                // Mempool updates are only done once we are synced.
+                if self.consensus.is_ready_for_validation() {
+                    let block = self
+                        .consensus
+                        .blockchain
+                        .read()
+                        .get_block(&hash, true)
+                        .expect("Head block not found");
+
+                    self.mempool
+                        .update(&vec![(hash.clone(), block)], [].as_ref());
+                }
+            }
+            BlockchainEvent::Rebranched(old_chain, new_chain) => {
+                // Mempool updates are only done once we are synced.
+                if self.consensus.is_ready_for_validation() {
+                    self.mempool.update(&new_chain, &old_chain);
+                }
+            }
+            _ => {
+                // Nothing to do here.
+            }
+        }
+    }
+}
+
+impl<TValidatorNetwork: ValidatorNetwork> Stream for MempoolTask<TValidatorNetwork>
+where
+    PubsubId<TValidatorNetwork>: std::fmt::Debug + Unpin,
+{
+    type Item = MempoolEvent;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Process consensus updates.
+        // Start mempool as soon as we have consensus and can enforce the validity window.
+        // Stop the mempool if we lose consensus or cannot enforce the validity window.
+        while let Poll::Ready(Some(event)) = self.consensus_event_rx.poll_next_unpin(cx) {
+            match event {
+                Ok(ConsensusEvent::Established {
+                    synced_validity_window: true,
+                }) => self.init_mempool(),
+                Ok(ConsensusEvent::Lost)
+                | Ok(ConsensusEvent::Established {
+                    synced_validity_window: false,
+                }) => self.pause(),
+                Err(_) => return Poll::Ready(None),
+            }
+        }
+
+        // Process blockchain updates.
+        if let Poll::Ready(Some(event)) = self.blockchain_event_rx.poll_next_unpin(cx) {
+            trace!(?event, is_active = self.mempool_active, "blockchain event");
+            self.on_blockchain_event(&event);
+            return Poll::Ready(Some(MempoolEvent(event)));
+        }
+
+        Poll::Pending
+    }
+}
+
+impl<TValidatorNetwork: ValidatorNetwork> Future for MempoolTask<TValidatorNetwork>
+where
+    PubsubId<TValidatorNetwork>: std::fmt::Debug + Unpin,
```

Comment:
These should be put onto the `ValidatorNetwork` trait instead.

Reply:
Outdated.